drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.winsen-sensor.com/sensors/co2-sensor/mh-z19b.html
+ */
+
+#define DT_DRV_COMPAT winsen_mhz19b
+
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <drivers/sensor.h>
+#ifdef CONFIG_MHZ19B_KEEP_MCU_AWAKE
+#include <pm/pm.h>
+#endif
+
+#include <drivers/sensor/mhz19b.h>
+#include "mhz19b.h"
+
+LOG_MODULE_REGISTER(mhz19b, CONFIG_SENSOR_LOG_LEVEL);
+
+static void mhz19b_uart_flush(const struct device *uart_dev)
+{
+	uint8_t c;
+
+	while (uart_fifo_read(uart_dev, &c, 1) > 0) {
+		continue;
+	}
+}
+
+static uint8_t mhz19b_checksum(const uint8_t *data)
+{
+	uint8_t cs = 0;
+
+	for (uint8_t i = 1; i < MHZ19B_BUF_LEN - 1; i++) {
+		cs += data[i];
+	}
+
+	return 0xff - cs + 1;
+}
+
+static int mhz19b_send_cmd(const struct device *dev, const enum mhz19b_cmd_idx cmd_idx,
+			   const bool has_rsp)
+{
+	struct mhz19b_data *data = dev->data;
+	const struct mhz19b_cfg *cfg = dev->config;
+	int ret;
+
+	/* Make sure last command has been transferred */
+	ret = k_sem_take(&data->tx_sem, MHZ19B_WAIT);
+	if (ret) {
+		return ret;
+	}
+
+	data->cmd_idx = cmd_idx;
+	data->has_rsp = has_rsp;
+	k_sem_reset(&data->rx_sem);
+
+#ifdef CONFIG_MHZ19B_KEEP_MCU_AWAKE
+	if (has_rsp) {
+		pm_constraint_set(PM_STATE_SUSPEND_TO_IDLE);
+	}
+#endif
```

Comment:
This looks like a UART PM problem being passed to client drivers. This solution is not universal, some UART drivers may need another constraint (if they need it at all). Can you please describe the issue?

Reply:
This is for devices with uart drivers that don't support wake up from stop mode, i.e. stm32 uart drivers.

Reply:
It looks like it should be the UART driver that sets necessary constraints, not UART clients that can, in general, work with any UART. Not all UARTs will require the same constraints, so the highlighted code makes this driver not fully portable.

Reply:
It has to be done by the client.

The MCU sends command to the sensor device and wait for the device to reply. If the MCU goes to stop mode after sending the command, it will not be able receive the reply from the sensor device.

Reply:
IMO this is likely a serial API limitation. A generic device driver should rarely use `pm_constraint` API. For some UARTs setting `PM_STATE_SUSPEND_TO_IDLE` will not be sufficient, only the serial driver knows which states need to be constrained (if any).

Reply:
Agreed, as the modem drivers also do not account for this case.